util/src/trie/triedbmut.rs

```diff
@@ -259,259 +444,280 @@ impl<'db> TrieDBMut<'db> {
-			});
-			trace!("take_node {:?} (indirect for {:?})", rlp.as_raw().pretty(), r);
-			journal.delete_node_sha3(h);
-			r
-		}
-		else {
-			trace!("take_node {:?} (???)", rlp.as_raw().pretty());
-			panic!("Empty or invalid node given?");
-		}
+		self.inspect(stored, move |trie, node| trie.remove_inspector(node, partial))
+			.map(|(new, changed)| (self.storage.alloc(new), changed))
 	}
 
-	#[cfg_attr(feature="dev", allow(cyclomatic_complexity))]
-	/// Determine the RLP of the node, assuming we're inserting `partial` into the
-	/// node currently of data `old`. This will *not* delete any hash of `old` from the database;
-	/// it will just return the new RLP that includes the new node.
-	///
-	/// `journal` will record the database updates so as to make the returned RLP valid through inserting
-	/// and deleting nodes as necessary.
-	///
-	/// **This operation will not insert the new node nor destroy the original.**
-	fn augmented(&self, old: &[u8], partial: &NibbleSlice, value: &[u8], journal: &mut Journal) -> Bytes {
-		trace!("augmented (old: {:?}, partial: {:?}, value: {:?})", old.pretty(), partial, value.pretty());
-		// already have an extension. either fast_forward, cleve or transmute_to_branch.
-		let old_rlp = Rlp::new(old);
-		match old_rlp.prototype() {
-			Prototype::List(17) => {
-				trace!("branch: ROUTE,AUGMENT");
-				// already have a branch. route and augment.
-				let mut s = RlpStream::new_list(17);
-				let index = if partial.is_empty() {16} else {partial.at(0) as usize};
-				for i in 0..17 {
-					match index == i {
-						// not us - leave alone.
-						false => { s.append_raw(old_rlp.at(i).as_raw(), 1); },
-						// branch-leaf entry - just replace.
-						true if i == 16 => { s.append(&value); },
-						// original had empty slot - place a leaf there.
-						true if old_rlp.at(i).is_empty() => journal.new_node(Self::compose_leaf(&partial.mid(1), value), &mut s),
-						// original has something there already; augment.
-						true => {
-							let new = self.augmented(self.take_node(&old_rlp.at(i), journal), &partial.mid(1), value, journal);
-							journal.new_node(new, &mut s);
+	/// the removal inspector
+	fn remove_inspector(&mut self, node: Node, partial: NibbleSlice) -> Action {
+		match (node, partial.is_empty()) {
+			(Node::Empty, _) => Action::Delete,
+			(Node::Branch(c, None), true) => Action::Restore(Node::Branch(c, None)),
+			(Node::Branch(children, _), true) => {
+				// always replace since we took the value out.
+				Action::Replace(self.fix(Node::Branch(children, None)))
+			}
+			(Node::Branch(mut children, value), false) => {
+				let idx = partial.at(0) as usize;
+				if let Some(child) = children[idx].take() {
+					trace!(target: "trie", "removing value out of branch child, partial={:?}", partial);
+					match self.remove_at(child, partial.mid(1)) {
+						Some((new, changed)) => {
+							children[idx] = Some(new.into());
+							let branch = Node::Branch(children, value);
+							if changed {
```

Comment:
`} else {` line is a bit wasteful for these trivial ternary-like conditionals. typically use `match` instead.